### PR TITLE
Draft CorrelationId using Activity

### PR DESCRIPTION
**Pull request by msallin**
Let me know what you think. It will bring some hours of work so I will only continue if you agree that it would be a cool feature and we should have it implemented like the WebHost and the HttpClient

### CODE REVIEW

File: src/Silverback.Integration/Messaging/Broker/Consumer.cs [hunk 1]
```diff
@@ -1,59 +1,109 @@
...
-        protected new TBroker Broker => (TBroker)base.Broker;
-
-        protected new TEndpoint Endpoint => (TEndpoint)base.Endpoint;
-    }
-}
\ No newline at end of file
+﻿// Copyright (c) 2018-2019 Sergio Aquilini
+// This code is licensed under MIT license (see LICENSE file for details)
+
+using System;
```
Comment (msallin) [inline src/Silverback.Integration/Messaging/Broker/Consumer.cs]: Please ignore whitespace changes here.
Btw. which line ending do you use in this repo?

File: src/Silverback.Integration/Messaging/Broker/Consumer.cs [hunk 2]
```diff
@@ -1,59 +1,109 @@
...
+using System.Collections.Generic;
+using System.Diagnostics;
+using Microsoft.Extensions.Logging;
+using Silverback.Messaging.Messages;
+
+namespace Silverback.Messaging.Broker
+{
+    public abstract class Consumer : EndpointConnectedObject, IConsumer
```
Comment (msallin) [inline src/Silverback.Integration/Messaging/Broker/Consumer.cs]: I took the code from the HttpHost as sample to implement it here.

File: src/Silverback.Integration/Messaging/Broker/Consumer.cs [hunk 3]
```diff
@@ -1,59 +1,109 @@
...
+            _messageLogger.LogTrace(_logger, "Message received.", deserializedMessage, Endpoint);
+
+            Received.Invoke(this, new MessageReceivedEventArgs(deserializedMessage, headers, offset));
+
+            StopActivity(activity);
+        }
+
+        private Activity StartActivity(IEnumerable<MessageHeader> messageHeaders)
+        {
+            string correlationId =  messageHeaders.GetFromHeaders(CorrelationIdHeaderKey);
```
Comment (msallin) [inline src/Silverback.Integration/Messaging/Broker/Consumer.cs]: Here actually the magic happens.
If there is a correlationIn the header, its used as Parent.
If not then just a new Activity without a parentId is created.
From here on, the ActivityId flows through the system.

File: src/Silverback.Integration/Messaging/Broker/Consumer.cs (hunk 1 above)
Comment (BEagle1984) [reply src/Silverback.Integration/Messaging/Broker/Consumer.cs]: Uhm...regular crlf? I don't know, I didn't configure anything special (on purpose).

**Comment by BEagle1984** [comment]
@msallin the activity stuff it is definitely a good idea. A couple of remarks:
- I just would like to decouple it from the Consumer and maybe even outsource it into another package
- Some time ago I added the behaviors with exactly such kind of scenario in mind, have a look and see if we could move the magic from the consumer into a behavior! https://beagle1984.github.io/silverback/docs/quickstart/behaviors

Let me know what do you think. I'll be out a couple of days but I'll have a closer look as soon as I'm back (saturday or sunday).

Thanks for contributing!

**Comment by msallin** [comment]
Will reopen it as soon as I got the first working (and understandable) version.

**Pull request closed without merging**